docs/chapter_tree/avl_tree.md

```diff
@@ -0,0 +1,308 @@
+# AVL 树
+
+## AVL 树的起源
+
+在「二叉搜索树」章节中提到，如进行多次插入与删除操作后，二叉搜索树可能会退化为链表。此时所有操作的时间复杂度都会由 $O(\log n)$ 劣化至 $O(n)$ 。例如，删除结点 4 后，该二叉搜索树就会退化为链表。
+
+=== "删除前"
+    ![binary search tree1](avl_tree.assets/binary_search_tree1.png)
+=== "删除后"
+    ![binary search tree2](avl_tree.assets/binary_search_tree2.png)
+
+为了解决这一问题，G. M. Adelson-Velsky 和 E. M. Landis 在其 1962 年发表的论文 "An algorithm for the organization of info
+rmation" 中提出了「平衡二叉搜索树」，也以两位作者命名，常被称为「AVL 树」。
+
+## AVL 树的性质
+
+「AVL 树」既是「二叉搜索树」又是「平衡二叉树」，同时满足这两种二叉树的所有性质。「平衡二叉树」规定树中任意结点左右子树的高度差的绝对值不能超过 1 。本文定义：
+
+- 「平衡因子 Balance Factor」为 **左子树的高度减右子树的高度** 。
+- 空树的高度定义为 0 ，叶结点的高度定义为 1 。
+
+!!! tip
+
+    设「平衡因子」为 $f$ ，则一棵 AVL 树的任意结点的平衡因子满足 $-1 \le f \le 1$
+
+## AVL 树的优势
+
+提出 AVL 树的两位大佬的厉害之处在于，他们设计了一系列操作，使得 AVL 树在不断添加与删除结点后，仍然不会发生退化，进而使得各种操作的时间复杂度均能保持在 $O(\log n)$ 级别。
+
+## AVL 树的操作
+
+### 查找结点
+
+「 AVL 树」的结点查找操作与「二叉搜索树」一致，在此不再赘述。
+
+### 插入结点
+
+由于平衡二叉树需要保证其任意结点的平衡因子满足限制，所以在插入结点后可能会造成 AVL 树的失衡。例如，平衡二叉树在插入结点 0 前 / 后如下图所示（括号内表示当前结点的平衡因子）：
+
+=== "插入前"
+    ![avl tree 1](avl_tree.assets/avl_tree1.png)
+=== "插入后"
+    ![avl tree 2](avl_tree.assets/avl_tree2.png)
+
+观察发现，插入后结点 2 和 结点 3 已经不满足平衡二叉树的性质，我们将这一现象称为 **失衡** 。为了解决该问题，首先需要分析哪些结点会出现失衡，经过观察可得两条规律：
```

Comment:
I modify the description and the mathematical proof. Please check the correctness.